File: src/mcp/process.ts
import { canOperate, deviceRole, logEvent } from "../core";
import { id, now, q } from "../db";
import { markProcessLost, workspaceForDevice } from "../process-store";
import type { AgentClientMessage } from "../protocol";
import { sendMcpAgent } from "./relay";
import type { McpToolContext } from "./types";

const OUTPUT_LIMIT = 256 * 1024;
const COMPLETED_TTL = 5 * 60_000;
const ACTIVE_TTL = 30 * 60_000;
const MAX_STATES = 128;

export type McpProcessResult = {
  processId: string;
  status: "exited" | "running" | "lost";
  output: string;
  exitCode: number | null;
  signal: string | null;
  error: string | null;
  nextOffset: number;
  outputTruncated: boolean;
};

type Listener = () => void;
type State = {
  processId: string; grantId: string; userId: string; deviceId: string; status: McpProcessResult["status"];
  output: string; outputTruncated: boolean; exitCode: number | null; signal: string | null; error: string | null;
  createdAt: number; updatedAt: number; initial?: { resolve: (value: McpProcessResult) => void; timer: ReturnType<typeof setTimeout> };
  listeners: Set<Listener>;
};

const states = new Map<string, State>();

function appendOutput(state: State, chunk: string) {
  if (!chunk || state.output.length >= OUTPUT_LIMIT) {
    if (chunk) state.outputTruncated = true;
    return;
  }
  const room = OUTPUT_LIMIT - state.output.length;
  state.output += chunk.slice(0, room);
  if (chunk.length > room) state.outputTruncated = true;
}

function decodeProcessData(value: string) {
  try { return Buffer.from(value, "base64url").toString("utf8"); }
  catch { return ""; }
}

function result(state: State, offset = 0): McpProcessResult {
  const start = Math.min(Math.max(0, offset), state.output.length);
  return { processId: state.processId, status: state.status, output: state.output.slice(start), exitCode: state.exitCode,
    signal: state.signal, error: state.error, nextOffset: state.output.length, outputTruncated: state.outputTruncated };
}

function notify(state: State) {
  state.updatedAt = now();
  for (const listener of [...state.listeners]) listener();
}

function cleanupStates() {
  const t = now();
  for (const [processId, state] of states) {
    const ttl = state.status === "running" ? ACTIVE_TTL : COMPLETED_TTL;
    if (state.updatedAt + ttl < t) states.delete(processId);
  }
  if (states.size < MAX_STATES) return;
  const removable = [...states.values()].sort((a, b) => a.updatedAt - b.updatedAt);
  while (states.size >= MAX_STATES && removable.length) states.delete(removable.shift()!.processId);
}

function settleInitial(state: State) {
  if (!state.initial) return;
  const initial = state.initial; delete state.initial; clearTimeout(initial.timer); initial.resolve(result(state));
}

export function runMcpProcess(context: McpToolContext, input: {
  deviceId: string; command: string; cwd?: string; timeoutSeconds?: number;
}) {
  cleanupStates();
  const { grant, payload } = context, deviceId = String(input.deviceId || "");
  if (!payload.deviceIds.includes(deviceId)) throw new Error("device is outside this MCP grant");
  if (!canOperate(deviceRole(payload.userId, deviceId))) throw new Error("operator access is no longer available for this device");
  const command = String(input.command || "").trim(), cwd = String(input.cwd || "").trim().slice(0, 4096);
  if (!command || command.length > 8192) throw new Error("invalid command");
  const processId = id(), t = now();
  q(`INSERT INTO processes(id,device_id,origin,status,terminal,created_by,created_at)
    VALUES(?,?,'mcp','starting',0,?,?)`).run(processId, deviceId, payload.userId, t);
  logEvent("mcp.process.created", workspaceForDevice(deviceId), payload.userId, deviceId,
    { grantId: payload.id, client: payload.clientName, processId });
  const state: State = { processId, grantId: payload.id, userId: payload.userId, deviceId, status: "running", output: "",
    outputTruncated: false, exitCode: null, signal: null, error: null, createdAt: t, updatedAt: t, listeners: new Set() };
  states.set(processId, state);
  const timeout = Math.min(60, Math.max(1, Number(input.timeoutSeconds || 20))) * 1000;
  return new Promise<McpProcessResult>((resolve) => {
    state.initial = { resolve, timer: setTimeout(() => { delete state.initial; resolve(result(state)); }, timeout) };
    const sent = sendMcpAgent(deviceId, {
      type: "mcp.process.start", id: processId, command, cwd, userId: payload.userId,
      mcpGrant: grant.grant, mcpSignature: grant.grant_signature,
      grant: grant.control_grant, credentialId: grant.credential_id, assertion: grant.control_assertion,
    });
    if (!sent) {
      markMcpProcessLost(processId, "RC Node is offline"); markProcessLost(processId, "RC Node is offline");
    }
  });
}

export function markMcpProcessLost(processId: string, error: string) {
  const state = states.get(processId); if (!state || state.status !== "running") return;
  state.status = "lost"; state.error = String(error || "process lost").slice(0, 1024); settleInitial(state); notify(state);
}

export async function mcpProcessStatus(context: McpToolContext, processId: string, offset = 0, waitSeconds = 0) {
  cleanupStates();
  const state = states.get(processId);
  if (!state || state.grantId !== context.payload.id || state.userId !== context.payload.userId || !context.payload.deviceIds.includes(state.deviceId)) {
    throw new Error("process status is unavailable for this MCP grant");
  }
  const startOffset = Math.min(Math.max(0, Number(offset) || 0), state.output.length);
  const wait = Math.min(60, Math.max(0, Number(waitSeconds) || 0)) * 1000;
  if (!wait || state.status !== "running" || state.output.length > startOffset) return result(state, startOffset);
  return await new Promise<McpProcessResult>((resolve) => {
    const done = () => { clearTimeout(timer); state.listeners.delete(done); resolve(result(state, startOffset)); };
    const timer = setTimeout(done, wait);
    state.listeners.add(done);
  });
}

export function handleMcpProcessMessage(process: any, message: AgentClientMessage) {
  if (process?.origin !== "mcp") return false;
  const state = states.get(process.id);
  if (message.type === "process.stdout" || message.type === "process.stderr") {
    if (state) { appendOutput(state, decodeProcessData(message.data || "")); notify(state); }
    return true;
  }
  if (message.type === "process.exit") {
    if (state) {
      appendOutput(state, message.output || ""); state.status = "exited"; state.exitCode = message.exitCode ?? null;
      state.signal = message.signal || null; state.error = null; settleInitial(state); notify(state);
    }
    return false;
  }
  return message.type === "process.started";
}

File: crates/rc-server/src/mcp_tools/process.rs
use super::{McpContext, complete, require_owned_device};
use crate::{AppState, McpProcessResult, now_ms};
use rc_protocol::ServerToNode;
use uuid::Uuid;

mod node_status;
mod request;
use node_status::{lost_result, query_node_status};
use request::{environment, execution_mode};

pub(super) async fn run(
    state: &AppState,
    context: &McpContext,
    args: &serde_json::Map<String, serde_json::Value>,
) -> anyhow::Result<serde_json::Value> {
    let device = args
        .get("deviceId")
        .and_then(serde_json::Value::as_str)
        .unwrap_or_default();
    require_owned_device(state, context, device)?;
    require_execution_v2(state, device)?;
    let mode = execution_mode(args)?;
    let environment = environment(args)?;
    let cwd = args
        .get("cwd")
        .and_then(serde_json::Value::as_str)
        .unwrap_or_default()
        .to_owned();
    let wait = args
        .get("waitSeconds")
        .or_else(|| args.get("timeoutSeconds"))
        .and_then(serde_json::Value::as_u64)
        .unwrap_or(20)
        .min(60);
    let max_runtime_seconds = args
        .get("maxRuntimeSeconds")
        .and_then(serde_json::Value::as_u64);
    let process_id = Uuid::new_v4().to_string();
    insert_process(state, context, device, &process_id)?;
    let message = ServerToNode::McpStart {
        process_id: process_id.clone(),
        user_id: context.payload.user_id.clone(),
        mode,
        cwd,
        environment,
        max_runtime_seconds,
        mcp_grant: context.record.grant.clone(),
        mcp_signature: context.record.grant_signature.clone(),
        control_grant: context.record.control_grant.clone(),
        credential_id: context.record.credential_id.clone(),
        control_assertion: context.record.control_assertion.clone(),
    };
    if let Err(error) = state.nodes.send(device, &message).await {
        let reason = format!("RC command delivery failed: {error}");
        state.lose_hosted_process(device, &process_id, &reason);
        return complete_result(lost_result(process_id, reason));
    }
    let result = query_node_status(state, context, device, &process_id, 0, wait).await?;
    complete_result(result)
}

fn require_execution_v2(state: &AppState, device: &str) -> anyhow::Result<()> {
    let capabilities = state.db.with_connection(|db| {
        db.query_row(
            "SELECT capabilities FROM devices WHERE id=?",
            [device],
            |row| row.get::<_, String>(0),
        )
    })?;
    let capabilities = serde_json::from_str::<Vec<String>>(&capabilities).unwrap_or_default();
    if !capabilities.iter().any(|value| value == "execution-v2") {
        anyhow::bail!("Node upgrade required: execution-v2 is unavailable");
    }
    Ok(())
}

pub(super) async fn status(
    state: &AppState,
    context: &McpContext,
    args: &serde_json::Map<String, serde_json::Value>,
) -> anyhow::Result<serde_json::Value> {
    let process_id = args
        .get("processId")
        .and_then(serde_json::Value::as_str)
        .unwrap_or_default();
    let device = args
        .get("deviceId")
        .and_then(serde_json::Value::as_str)
        .unwrap_or_default();
    if process_id.is_empty() {
        anyhow::bail!("invalid process ID");
    }
    let cursor = args
        .get("cursor")
        .or_else(|| args.get("offset"))
        .and_then(serde_json::Value::as_u64)
        .unwrap_or(0) as usize;
    let wait = args
        .get("waitSeconds")
        .and_then(serde_json::Value::as_u64)
        .unwrap_or(0)
        .min(60);
    require_owned_device(state, context, device)?;
    let result = query_node_status(state, context, device, process_id, cursor, wait).await?;
    complete_result(result)
}

fn insert_process(
    state: &AppState,
    context: &McpContext,
    device: &str,
    process_id: &str,
) -> anyhow::Result<()> {
    state.db.with_connection(|db| {
        db.execute(
            "INSERT INTO processes(id,device_id,origin,status,terminal,created_by,created_at) \
             VALUES(?,?,'mcp','starting',0,?,?)",
            rusqlite::params![process_id, device, context.payload.user_id, now_ms()],
        )?;
        Ok(())
    })?;
    Ok(())
}

fn complete_result(result: McpProcessResult) -> anyhow::Result<serde_json::Value> {
    let mut text = String::new();
    for chunk in &result.chunks {
        if chunk.stream == "stderr" {
            text.push_str("[stderr] ");
        }
        if chunk.encoding == "text" {
            text.push_str(&chunk.data);
        } else {
            text.push_str(&format!("[base64 output: {}]", chunk.data));
        }
    }
    if result.truncated_before_cursor > 0 {
        text.push_str(&format!(
            "\n[Earlier output before cursor {} is no longer buffered.]",
            result.truncated_before_cursor
        ));
    }
    let status = match result.status.as_str() {
        "exited" => format!(
            "Exit {}.",
            result
                .exit_code
                .map(|value| value.to_string())
                .unwrap_or_else(|| "unknown".into())
        ),
        "running" => format!("Process {} is still running.", result.process_id),
        _ => format!(
            "Process was lost{}",
            result
                .error
                .as_ref()
                .map(|error| format!(": {error}"))
                .unwrap_or_else(|| ".".into())
        ),
    };
    if !text.is_empty() && !text.ends_with('\n') {
        text.push('\n');
    }
    text.push_str(&status);
    if result.output_pending {
        text.push_str(" More buffered output is available at nextCursor.");
    }
    Ok(complete(serde_json::to_value(&result)?, text, false))
}
